rust/mlt/src/error.rs

```diff
@@ -1,23 +1,337 @@
+use bytes_varint::VarIntError as BvVarIntError;
+use cxx;
 use thiserror::Error;
 
+use crate::metadata::proto_tileset::ScalarType;
+use crate::metadata::stream_encoding::{LogicalLevelTechnique, PhysicalLevelTechnique};
 pub type MltResult<T> = Result<T, MltError>;
 
 #[derive(Error, Debug)]
+#[non_exhaustive]
 pub enum MltError {
     #[error("Unable to parse property: {0}")]
     PropertyParseError(String),
     #[error("Unsupported key value type: {0}")]
     UnsupportedKeyType(String),
-    #[error("Failed to read file: {0}")]
-    FileReadError(#[from] std::io::Error),
+    // #[error("Failed to read file: {0}")]
+    // FileReadError(#[from] std::io::Error),
     #[error("Unsupported geometry type: {0}")]
     UnsupportedGeometryType(String),
     #[error("Failed to decode protobuf: {0}")]
     DecodeError(String),
     #[error("Failed to decode metadata: {0}")]
     MetadataDecodeError(String),
-    #[error("Failed to decode rle: {0}")]
-    RleDecodeError(#[from] serde_columnar::ColumnarError),
+    // #[error("Failed to decode rle: {0}")]
+    // RleDecodeError(#[from] serde_columnar::ColumnarError),
     #[error("Unsupported technique in decode_int_stream: {0:?}")]
     UnsupportedIntStreamTechnique(String),
+
+    //---------------------------------------------------------
+    // Structured, allocation-light
+    //---------------------------------------------------------
+
+    //---------------------------------------------------------
+    // Parsing / validation
+    //---------------------------------------------------------
+    #[error("Property parse error: field={field} at={at:?} kind={kind:?}")]
+    PropertyParse {
+        field: &'static str,
+        at: Option<usize>,
+        kind: PropertyParseKind,
+    },
+
+    #[error("Unsupported key type: code={code}")]
+    UnsupportedKeyTypeCode { code: u8 },
```

Comment:
most of the simple errors should probably use tuple rather than struct pattern


```suggestion
    UnsupportedKeyTypeCode(u8),
```